Refactor S3 API to static factory methods

2m opened the pull request:
This is a second attempt to clean up S3 connector API. Instead of creating instances of `S3Client` that then are used to create `Source/Flow/Sink`s, all of the factories are moved to the `S3` object. If connector implementation needs an ActorSystem, ExecutionContext or S3Settings, all of those are taken from the materializer.

Access to the materialized is factored out to `MaterializerAccess`.

This PR also enables `MinioS3IntegrationSpec` and configures `minio` docker image for that test.

Continuation of #1323

Two test cases are failing. Therefore this is still WIP. One testcase `succeed uploading an empty file` is tricky to fix with the current implementation of `MaterializerAccess.source`, as empty file (empty source) does not work as other sources with `Source.lazily`.

ennru commented on s3/src/main/scala/akka/stream/alpakka/s3/S3Settings.scala:
```diff
@@ -67,7 +69,7 @@ object S3Settings {
    * Scala API: Creates [[S3Settings]] from a [[com.typesafe.config.Config Config]] object.
    */
-  def apply(config: Config, configurationPrefix: String = "akka.stream.alpakka.s3"): S3Settings = {
+  def apply(config: Config, configurationPrefix: String = configPath): S3Settings = {
```
With this rewrite, the `config.getConfig` should be removed and always read directly from there.

ennru commented on s3/src/main/scala/akka/stream/alpakka/s3/auth/Signer.scala:
```diff
@@ -7,28 +7,31 @@ package akka.stream.alpakka.s3.auth
-import akka.stream.Materializer
+import akka.stream.scaladsl.Source
 
 private[alpakka] object Signer {
```
`@InternalApi`
Looks like the whole package should go in `impl`.

ennru commented on s3/src/main/scala/akka/stream/alpakka/s3/auth/package.scala:
```diff
@@ -6,11 +6,10 @@ package akka.stream.alpakka.s3
 
-import scala.concurrent.Future
-
 package object auth {
```
`@InternalApi`

ennru commented on s3/src/main/scala/akka/stream/alpakka/s3/impl/MaterializerAccess.scala:
```diff
@@ -0,0 +1,74 @@
+
+  def sink[T, M](factory: Attributes => Sink[T, M]): Sink[T, Future[M]] =
+    Sink
+      .lazyInit(
```
Why not `Async`?

ennru commented on s3/src/main/scala/akka/stream/alpakka/s3/impl/MaterializerAccess.scala:
```diff
@@ -0,0 +1,74 @@
+
+import scala.concurrent.Future
+
+object MaterializerAccess {
```
`@InternalApi`

ennru commented on s3/src/main/scala/akka/stream/alpakka/s3/impl/HttpRequests.scala:
```diff
@@ -89,7 +93,7 @@ private[alpakka] object HttpRequests {
 
   def completeMultipartUploadRequest(upload: MultipartUpload, parts: Seq[(Int, String)])(
-      implicit ec: ExecutionContext,
+      implicit mat: ExecutionContext,
```
Strange rename?!

2m commented:
Addresses PR feedback. Also updated the way Materializer and Attributes are exposed. Instead of using `lazy{Source,Flow,Sink}` factories, `Setup.{source,flow,sink}` factories are introduced which wrap given graph into a custom GraphStage.

Materialized value is still wrapped into a `Future`, but the problem with empty sources, which `lazySink` had, is gone. 

Still needs a docs rewrite.

2m commented:
Docs have been refactored. The IBM Bluemix section has been moved to a separate page. Also added a new paragraph and code snipped about using S3Atributes.

ennru commented on docs/src/main/paradox/release-notes/1.0-M2.md:
```diff
@@ -0,0 +1,46 @@
+New API frees the factory methods from these dependencies.
+All of the required resources are injected during the materialization time.
+
+If you are loading `S3Settings` from the `application.conf` then migration should be done by changing all method calls on `s3Client` instance to method calls on the @scala[@scaladoc[S3](akka.stream.alpakka.s3.scaladsl.S3$)]@java[@scaladoc[S3](akka.stream.alpakka.s3.javadsl.S3$)] factory.
```
Mention that the settings live in `alpakka.s3` now.

ennru commented:
Fixes #1086

ennru commented:
Fixes #938

2m commented on s3/src/main/resources/reference.conf:
```diff
@@ -14,33 +14,22 @@ akka.stream.alpakka.s3 {
-    #   access-key-id = ${old-access-key-id}
-    #   secret-access-key = ${old-secret-access-key}
-    # }
     # If this section is absent, the fallback behavior is to use the
```
This comment describes what is a fallback behavior if complete `credentials { ... }` section is missing. I have changed it to refer to `provider = default` instead of repeating the same.

2m commented on s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala:
```diff
@@ -51,71 +52,60 @@ final case class ListBucketResult(isTruncated: Boolean,
-}
-
-private[alpakka] final class S3Stream(settings: S3Settings)(implicit system: ActorSystem, mat: Materializer) {
+private[alpakka] object S3Stream {
```
Fixed.

2m commented:
Addressed feedback and rebased on top of the current master.
